dropshot/src/schema_util.rs

```diff
@@ -116,25 +138,110 @@ pub(crate) fn schema2struct(
+                        ))
+                    }
                 }
             }
 
-            results
+            Ok(results)
         }
 
+        // Unit enums that do not have doc comments are presented as enum_values.
+        schemars::schema::Schema::Object(schemars::schema::SchemaObject {
+            metadata: _,
+            instance_type: _,
+            format: None,
+            enum_values: Some(_),
+            const_value: None,
+            subschemas: None,
+            number: None,
+            string: None,
+            array: None,
+            object: None,
+            reference: None,
+            extensions: _,
+        }) => Err(Box::new(Schema2StructError::UnitEnum(schema.clone()))),
+
+        // Complex enums, and unit enums that have doc comments, are presented
+        // as subschemas with instance_type and reference set to None.
+        schemars::schema::Schema::Object(schemars::schema::SchemaObject {
+            metadata: _,
+            instance_type: None,
+            format: None,
+            enum_values: None,
+            const_value: None,
+            subschemas: Some(_),
+            number: None,
+            string: None,
+            array: None,
+            object: None,
+            reference: None,
+            extensions: _,
+        }) => Err(Box::new(Schema2StructError::ComplexEnum(schema.clone()))),
+
         // The generated schema should be an object.
-        invalid => panic!("invalid type {:#?}", invalid),
+        invalid => {
+            Err(Box::new(Schema2StructError::InvalidType(invalid.clone())))
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum Schema2StructError {
+    InvalidType(schemars::schema::Schema),
+    InvalidSubschema(schemars::schema::SubschemaValidation),
+    UnitEnum(schemars::schema::Schema),
+    ComplexEnum(schemars::schema::Schema),
+}
+
+impl fmt::Display for Schema2StructError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Schema2StructError::InvalidType(schema) => {
+                write!(f, "invalid type {:#?}", schema)
+            }
+            Schema2StructError::InvalidSubschema(subschemas) => {
+                write!(f, "invalid subschema {:#?}", subschemas)
+            }
+            Schema2StructError::UnitEnum(schema) => {
+                write!(
+                    f,
+                    "invalid type {:#?}\n\
+                     (hint: this appears to be an enum with unit variants, \
+                     and needs to be wrapped in a struct)",
+                    schema
+                )
+            }
+            Schema2StructError::ComplexEnum(schema) => {
+                write!(
+                    f,
+                    "invalid type {:#?}\n\
+                     (hint: this appears to be an enum -- \
+                     if your enum only has unit variants, \
+                     it can be used in this position but needs to be wrapped in a struct)",
```

Comment:
is it important to produce distinct errors here? Would it be sufficient to say "put your enum in a struct, dummy!" for both cases?

Reply:
I do like that you're picking them apart... but I think the advice to the user is going to be basically the same.

Reply:
It looks like all of the non-simple schemas look exactly the same. I combined them anyway, since as you pointed out the advice is the same. (It also simplifies the test somewhat.)